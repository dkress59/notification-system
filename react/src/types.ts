import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react'

import { SpawnBannerProps, SpawnModalProps, SpawnToastProps } from ':core/types'

type CustomHTMLAttributes<
	T,
	K = void | Record<string, unknown>,
> = DetailedHTMLProps<HTMLAttributes<T> & K, T>

export type BannerAreaAttributes = CustomHTMLAttributes<HTMLBannerAreaElement>

export type NotificationAreaAttributes =
	CustomHTMLAttributes<HTMLNotificationAreaElement>

export type BannerNotificationAttributes =
	CustomHTMLAttributes<HTMLBannerNotificationElement>

export type ModalNotificationAttributes =
	CustomHTMLAttributes<HTMLModalNotificationElement>

export type ToastNotificationAttributes =
	CustomHTMLAttributes<HTMLToastNotificationElement>

declare global {
	// eslint-disable-next-line @typescript-eslint/no-namespace
	namespace JSX {
		interface IntrinsicElements {
			['banner-area']: BannerAreaAttributes
			['notification-area']: NotificationAreaAttributes
			['banner-notification']: BannerNotificationAttributes
			['modal-notification']: ModalNotificationAttributes
			['toast-notification']: ToastNotificationAttributes
		}
	}
}

export type BannerAreaProps = PropsWithChildren

export type NotificationAreaProps = PropsWithChildren

export interface BannerNotificationProps
	extends PropsWithChildren,
		Omit<SpawnBannerProps, 'content'> {
	onDismiss?: () => void
}

export interface ToastNotificationProps
	extends PropsWithChildren,
		Omit<SpawnToastProps, 'content'> {
	onDismiss?: () => void
}

export interface ModalNotificationProps
	extends PropsWithChildren,
		Omit<SpawnModalProps, 'content'> {
	onConfirm?: () => void
	onDecline?: () => void
	onDismiss?: () => void
}
