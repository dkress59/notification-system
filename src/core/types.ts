import {
	HTMLBannerAreaElement,
	HTMLBannerNotificationElement,
	HTMLModalNotificationElement,
	HTMLNotificationAreaElement,
	HTMLToastNotificationElement,
} from './components'

type SnakeToCamelCaseProp<S extends string> = S extends
	| `${infer T} ${infer U}`
	| `${infer T}-${infer U}`
	? `${Uncapitalize<T>}${Capitalize<SnakeToCamelCaseProp<U>>}`
	: Uncapitalize<S>
type SnakeToCamelCase<T> = {
	[K in keyof T as SnakeToCamelCaseProp<K & string>]: T[K]
}

export enum NotificationType {
	SUCCESS = 'success',
	INFO = 'info',
	ERROR = 'error',
	WARNING = 'warning',
}

export enum NotificationEvent {
	BANNER_DISMISSED = 'bannerDismissed',
	MODAL_CONFIRMED = 'modalConfirmed',
	MODAL_DECLINED = 'modalDeclined',
	MODAL_DISMISSED = 'modalDismissed',
	TOAST_DISMISSED = 'toastDismissed',
}

type SpawnArgs<T> = T & {
	content: string
}

export interface BannerAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnBannerArgs = SpawnArgs<SnakeToCamelCase<BannerAttributes>>

export interface ModalAttributes {
	'condition'?: boolean
	'headline'?: string
	'show-confirm'?: boolean
	'show-decline'?: boolean
	'type'?: NotificationType
	'label-confirm'?: string
	'label-decline'?: string
}
export type SpawnModalArgs = SpawnArgs<SnakeToCamelCase<ModalAttributes>>

export interface ToastAttributes {
	'auto-hide'?: boolean | string
	'auto-hide-after-ms'?: number | string
	'headline'?: string
	'type'?: NotificationType
}
export type SpawnToastArgs = SpawnArgs<SnakeToCamelCase<ToastAttributes>>

declare global {
	interface HTMLElementTagNameMap {
		'banner-area': HTMLBannerAreaElement
		'banner-notification': HTMLBannerNotificationElement
		'modal-notification': HTMLModalNotificationElement
		'notification-area': HTMLNotificationAreaElement
		'toast-notification': HTMLToastNotificationElement
	}
}
