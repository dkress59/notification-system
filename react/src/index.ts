export * from './components/banner-area'
export * from './components/banner-notification'
export * from './components/modal-notification'
export * from './components/notification-area'
export * from './components/toast-notification'
export * from './types'
