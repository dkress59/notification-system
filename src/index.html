<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
  <title>Notification Web Components</title>

  <script type="module" src="/build/notification-system.esm.js"></script>
  <script nomodule src="/build/notification-system.js"></script>

  <style>
    * {
      box-sizing: border-box;
    }

    body {
      font-family: sans-serif;
    }

    #notification-container {
      position: fixed;
      z-index: 50;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      pointer-events: none;
      display: flex;
      flex-direction: column;
      align-items: flex-end;
      overflow-y: auto;
    }

    /* #notification-container > * {
        right: 0;
      } */
  </style>
  <script>
    document.addEventListener('DOMContentLoaded', () => {
      document.querySelectorAll('notify-toast').forEach(toast => {
        toast.addEventListener('toastDismissed', (shadowRoot) => console.debug(
          'toast dismissed!',
          { shadowRoot }
        ))
      })
    })
  </script>
</head>

<body>
  <div id="notification-container">
    <!-- <my-component first="Stencil" last="'Don't call me a framework' JS"></my-component>
      <my-component first="more Stencil" middle="please" last="'Don't call me a framework' JS"></my-component> -->
    <notify-toast type="error" headline="Lorem Ipsum">Lorem ipsum dolor sit amet consectetur adipisicing elit, facilis
      necessitatibus sunt quisquam officia quidem.</notify-toast>
    <notify-toast>Mock Content</notify-toast>
    <notify-toast auto-hide>Mock Content</notify-toast>
    <notify-toast type="info" is-hidden>Hidden Content</notify-toast>
    <notify-toast>Mock Content</notify-toast>
  </div>
</body>

</html>