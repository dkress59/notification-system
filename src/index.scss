@import 'css/base.scss';
// @import 'css/util.scss';
