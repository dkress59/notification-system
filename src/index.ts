export { Components, JSX } from './components'
export { BannerArea } from './components/banner-area/banner-area'
export { BannerNotification } from './components/banner-notification/banner-notification'
export { ModalNotification } from './components/modal-notification/modal-notification'
export { NotificationArea } from './components/notification-area/notification-area'
export { ToastNotification } from './components/toast-notification/toast-notification'
