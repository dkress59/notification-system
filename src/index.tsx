import './index.scss'

import React from 'react'
import ReactDOM from 'react-dom'

import { App } from './app'
import { NotificationProvider } from './hooks'

ReactDOM.render(
	<React.StrictMode>
		<NotificationProvider>
			<App />
		</NotificationProvider>
	</React.StrictMode>,
	document.getElementById('app'),
)
