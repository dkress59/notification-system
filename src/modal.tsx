import React, { useEffect, useState } from 'react'

import { FinalModalProps, NotificationType } from './types'
import { getIcon, getModalClassName } from './util'

export function Modal({
	children,
	removeThisFromDom,
	labelAccept,
	labelDecline,
	onAccept,
	onDecline,
	title,
	type = NotificationType.SUCCESS,
}: FinalModalProps) {
	const hiddenClassName = 'opacity-0 mt-8 scale-95'
	const visibleClassName = 'opacity-100 mt-0 scale-100'

	const [transition, setTransition] = useState(hiddenClassName)

	const className = getModalClassName(type) + ' ' + transition
	const onDismiss = () => {
		if (onDecline) onDecline()
		setTransition(hiddenClassName)
	}

	useEffect(() => {
		setTransition(visibleClassName)
	}, [])

	useEffect(() => {
		const timeOut =
			transition === hiddenClassName
				? setTimeout(() => removeThisFromDom(), 500)
				: null
		return () => {
			if (timeOut) clearTimeout(timeOut)
		}
	}, [transition])

	return (
		<aside {...{ className }} data-testid="modal-component">
			{title && <h4>{title}</h4>}
			<span className="icon" data-testid="modal-icon">
				{getIcon(type)}
			</span>
			<div>{children}</div>
			<span
				className="dismiss"
				data-testid="dismiss-button"
				onClick={onDismiss}
			>
				&#x2715;
			</span>
			{onAccept && (
				<button onClick={onAccept} data-testid="accept-button">
					{labelAccept ?? 'Accept'}
				</button>
			)}
			{onDecline && (
				<button onClick={onDismiss} data-testid="decline-button">
					{labelDecline ?? 'Decline'}
				</button>
			)}
		</aside>
	)
}
