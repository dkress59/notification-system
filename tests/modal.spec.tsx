import { act, fireEvent, render, screen } from '@testing-library/react'

import { Modal } from '../src/modal'
import { FinalModalProps, NotificationType } from '../src/types'

const mockProps: FinalModalProps = {
	children: <>Mock Content</>,
	condition: true,
	labelConfirm: 'Mock Accept Label',
	labelDecline: 'Mock Decline Label',
	onConfirm: jest.fn(),
	onDecline: jest.fn(),
	removeThisFromDom: jest.fn(),
	title: 'Mock Title',
	type: NotificationType.SUCCESS,
}

describe('<Modal />', () => {
	it('initialises to correct defaults', () => {
		const { container } = render(<Modal removeThisFromDom={jest.fn()} />)
		expect(container).toMatchSnapshot()
	})
	it('initialises to correct default labels', () => {
		const { container } = render(
			<Modal
				removeThisFromDom={jest.fn()}
				onConfirm={jest.fn()}
				onDecline={jest.fn()}
			/>,
		)
		expect(container).toMatchSnapshot()
	})
	it('type=success applies correct className', () => {
		render(<Modal {...mockProps} />)
		const modal = screen.getByTestId('modal-component')
		expect(modal.classList).toContain('custom-modal-type-success')
	})
	it('type=info applies correct className', () => {
		render(<Modal {...mockProps} type={NotificationType.INFO} />)
		const modal = screen.getByTestId('modal-component')
		expect(modal.classList).toContain('custom-modal-type-info')
	})
	it('type=warning applies correct className', () => {
		render(<Modal {...mockProps} type={NotificationType.WARNING} />)
		const modal = screen.getByTestId('modal-component')
		expect(modal.classList).toContain('custom-modal-type-warning')
	})
	it('type=error displays the correct icon for', () => {
		render(<Modal {...mockProps} type={NotificationType.ERROR} />)
		const icon = screen.getByTestId('modal-icon')
		expect(icon.innerHTML).toBe(
			'<span class="text-xl inline-block -mt-1 transform-gpu -rotate-12">â•²</span>',
		)
	})
	it('fades out on dismissal', async () => {
		render(<Modal {...mockProps} type={NotificationType.ERROR} />)
		const modal = screen.getByTestId('modal-component')
		const dismissButton = modal.querySelector('.dismiss')!
		expect(modal.classList).toContain('opacity-100')
		fireEvent.click(dismissButton)
		await act(() => Promise.resolve())
		expect(modal.classList).toContain('opacity-0')
	})
	it('removes itself from DOM after fading out', async () => {
		const mockRemoveFromDom = jest.fn()
		render(<Modal {...mockProps} removeThisFromDom={mockRemoveFromDom} />)
		const modal = screen.getByTestId('modal-component')
		const dismissButton = modal.querySelector('.dismiss')!
		fireEvent.click(dismissButton)
		await act(() => Promise.resolve())
		expect(mockRemoveFromDom).not.toHaveBeenCalled()
		await act(
			() => new Promise(resolve => setTimeout(() => resolve(), 600)),
		)
		expect(mockRemoveFromDom).toHaveBeenCalled()
	})
	it('dismisses itself onConfirm', async () => {
		const mockRemoveFromDom = jest.fn()
		const mockOnConfirm = jest.fn()
		render(
			<Modal
				{...mockProps}
				removeThisFromDom={mockRemoveFromDom}
				onConfirm={mockOnConfirm}
			/>,
		)
		const confirmButton = screen.getByTestId('confirm-button')!
		fireEvent.click(confirmButton)
		await act(() => Promise.resolve())
		expect(mockOnConfirm).toHaveBeenCalled()
		await act(
			() => new Promise(resolve => setTimeout(() => resolve(), 600)),
		)
		expect(mockRemoveFromDom).toHaveBeenCalled()
	})
	it('dismisses itself onDecline', async () => {
		const mockRemoveFromDom = jest.fn()
		const mockOnDecline = jest.fn()
		render(
			<Modal
				{...mockProps}
				removeThisFromDom={mockRemoveFromDom}
				onDecline={mockOnDecline}
			/>,
		)
		const declineButton = screen.getByTestId('decline-button')!
		fireEvent.click(declineButton)
		await act(() => Promise.resolve())
		expect(mockOnDecline).toHaveBeenCalled()
		await act(
			() => new Promise(resolve => setTimeout(() => resolve(), 600)),
		)
		expect(mockRemoveFromDom).toHaveBeenCalled()
	})
	it('does not allow onAccept if condition=false', async () => {
		const mockRemoveFromDom = jest.fn()
		const mockOnConfirm = jest.fn()
		render(
			<Modal
				{...mockProps}
				removeThisFromDom={mockRemoveFromDom}
				onConfirm={mockOnConfirm}
				condition={false}
			/>,
		)
		const confirmButton = screen.getByTestId('confirm-button')!
		fireEvent.click(confirmButton)
		await act(() => Promise.resolve())
		expect(mockOnConfirm).not.toHaveBeenCalled()
		await act(
			() => new Promise(resolve => setTimeout(() => resolve(), 600)),
		)
		expect(mockRemoveFromDom).not.toHaveBeenCalled()
	})
})
