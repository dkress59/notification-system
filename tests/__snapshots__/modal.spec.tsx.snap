// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> initialises to correct default labels 1`] = `
<div>
  <aside
    class="custom-modal custom-modal-type-success opacity-100 mt-0 scale-100"
    data-testid="modal-component"
  >
    <span
      class="icon"
      data-testid="modal-icon"
    >
      ✓
    </span>
    <div />
    <span
      class="dismiss"
      data-testid="dismiss-button"
    >
      ✕
    </span>
    <button
      data-testid="accept-button"
    >
      Accept
    </button>
    <button
      data-testid="decline-button"
    >
      Decline
    </button>
  </aside>
</div>
`;

exports[`<Modal /> initialises to correct defaults 1`] = `
<div>
  <aside
    class="custom-modal custom-modal-type-success opacity-100 mt-0 scale-100"
    data-testid="modal-component"
  >
    <span
      class="icon"
      data-testid="modal-icon"
    >
      ✓
    </span>
    <div />
    <span
      class="dismiss"
      data-testid="dismiss-button"
    >
      ✕
    </span>
  </aside>
</div>
`;
